In [11]:
import tensorflow as tf
import numpy as np
print(tf.__version__)

1.15.0-rc3


## Variable and get_variable():

**There are two ways of creating a Variable in graph:**


1) `tf.get_variable()` method

2) `tf.Variable()` class

In [0]:
tf.reset_default_graph()
# There are two ways of create Variable in Graph:

# FIRST:
# init = tf.constant(4.)
# x = tf.get_variable('x', initializer=init)

# SECOND:
x = tf.Variable(5.0)

f = x ** 2

### Two key difference between `Variable` and `get_variable`:
1)  `tf.Variable` will always create a new variable, whereas `tf.get_variable` gets an existing variable with specified parameters from the graph, and if it doesn't exist, creates a new one.

2) `tf.Variable` requires that an initial value be specified.

### Optimization: we will optimize function $y=x^2$

In [0]:
optimizer = tf.train.GradientDescentOptimizer(0.1)
step = optimizer.minimize(f, var_list=[x])

In [100]:
tf.trainable_variables()

[<tf.Variable 'Variable:0' shape=() dtype=float32_ref>]

In [105]:
with tf.Session() as s:  # in this way session will be closed automatically
    s.run(tf.global_variables_initializer())
    for i in range(10):
        curr_x, curr_f = s.run([x, f])
        _ = s.run([step])
        print('Iteration: ', i)
        print('x = ', curr_x, '|', 'f = ', curr_f)
        print('---------------')

Iteration:  0
x =  5.0 | f =  25.0
---------------
Iteration:  1
x =  4.0 | f =  16.0
---------------
Iteration:  2
x =  3.2 | f =  10.240001
---------------
Iteration:  3
x =  2.56 | f =  6.5536
---------------
Iteration:  4
x =  2.0479999 | f =  4.1943035
---------------
Iteration:  5
x =  1.6383998 | f =  2.684354
---------------
Iteration:  6
x =  1.3107198 | f =  1.7179865
---------------
Iteration:  7
x =  1.0485759 | f =  1.0995114
---------------
Iteration:  8
x =  0.8388607 | f =  0.70368725
---------------
Iteration:  9
x =  0.6710886 | f =  0.45035988
---------------


### Just in order to ckeck manually calculations:
1. $x_0 = 5$
2. $x_1 = x_0 - lr \cdot f'(x_0) = 5 - 0.1 \cdot 10 = 4$
3. $x_2 = x_1 - lr \cdot f'(x_1) = 4 - 0.1 \cdot 8 = 3.2$
4. $x_3 = x_2 - lr \cdot f'(x_2) = 3.2 - 0.1 \cdot 6.4 = 2.56$
5. ...

### One more example for for sophisticated function:
$f(x)=2x^5−5x^4+x^3+3x^2−x$
Here is the graph: https://www.desmos.com/calculator/i3rnu6mrsy

This function has 2 local minum at points approximately:
$x_0=0.167$, $x_0=1.604$ and also $\lim \limits_{x \to -\infty}^{} f(x)= - \infty$

You can try several start points and different learning rates. Eventually you will reach one of those three points, i.e $x_0=0.167$, $x_0=1.604$ or $x_0 = - \infty$

In [0]:
tf.reset_default_graph()
# There are two ways of create Variable in Graph:

# FIRST:
# init = tf.constant(4.)
# x = tf.get_variable('x', initializer=init)

# SECOND:
x = tf.Variable(5.0)


f = 2 * x ** 5 - 5 * x ** 4 + x ** 3 + 3 * x ** 2 - x

In [0]:
optimizer = tf.train.GradientDescentOptimizer(0.001)
step = optimizer.minimize(f, var_list=[x])

In [117]:
tf.trainable_variables()

[<tf.Variable 'Variable:0' shape=() dtype=float32_ref>]

In [119]:
with tf.Session() as s:  # in this way session will be closed automatically
    s.run(tf.global_variables_initializer())
    for i in range(300):
        curr_x, curr_f = s.run([x, f])
        _ = s.run([step])
        print('Iteration: ', i)
        print('x = ', curr_x, '|', 'f = ', curr_f)
        print('---------------')

Iteration:  0
x =  5.0 | f =  3320.0
---------------
Iteration:  1
x =  1.1459998 | f =  -0.37174642
---------------
Iteration:  2
x =  1.149037 | f =  -0.38099778
---------------
Iteration:  3
x =  1.1520915 | f =  -0.39035416
---------------
Iteration:  4
x =  1.1551632 | f =  -0.39981544
---------------
Iteration:  5
x =  1.1582518 | f =  -0.4093809
---------------
Iteration:  6
x =  1.1613572 | f =  -0.41904998
---------------
Iteration:  7
x =  1.1644791 | f =  -0.42882192
---------------
Iteration:  8
x =  1.1676174 | f =  -0.4386965
---------------
Iteration:  9
x =  1.170772 | f =  -0.4486736
---------------
Iteration:  10
x =  1.1739426 | f =  -0.45875168
---------------
Iteration:  11
x =  1.1771289 | f =  -0.46892846
---------------
Iteration:  12
x =  1.1803309 | f =  -0.47920573
---------------
Iteration:  13
x =  1.1835481 | f =  -0.48958075
---------------
Iteration:  14
x =  1.1867803 | f =  -0.5000514
---------------
Iteration:  15
x =  1.1900275 | f =  -0.5106199
----

Notice that not always you will reach the closest minimum.
For example, take initial_value = 4., lr = 0.1

The closest minimum is obviously $x_0=1.604$
However, at first iteration you will achive that:
$x_{new} = x_0 - lr \cdot f'(x_0) = 4 - 0.01 \cdot 1351.0 = -9.51$
And after it you will reach $-\infty$